### PR TITLE
add TypeScript version

### PR DESCRIPTION
**Pull request by ilandikov**
Hi @emilybache 

I've recently seen your [video](https://www.youtube.com/watch?v=4bCuxMhtf_E) (thanks to @claremacrae!).

I understood that the purpose of this exercise is not to develop class behaviour (or data behaviour in other words), but rather developing an abstraction/interface of how classes should interact (by sending data between them). I realised I was doing that previously in a rather intuitive manner, which basically means I have no idea how I have been doing that =) And given that recently I had a similar task a I was struggling with, I decided to translate this kata to TypeScript which is my main language and do it for myself. So here we are =)

Thanks a lot for this kata! =)

### CODE REVIEW

**Comment by ilandikov** [comment]
I would be grateful if you could comment my solution: https://github.com/ilandikov/Auction-TestDesign-Kata/compare/typescript...ilandikov:Auction-TestDesign-Kata:typescript-solution?expand=1

Especially this particular commit: https://github.com/ilandikov/Auction-TestDesign-Kata/commit/27044fbb11f6cbb54a0e3cb76a518666ee4b278a

I know that in the video you were talking about a GOOS-style programming "Tell, don't ask" and having a ` expect(mockListener.actionCloser).toHaveBeenCalledTimes(1);` look probably like "asking". Still, I'm curious about your opinion on this. It just seems weird to me to write `checkExpectations()` while jest provides an easy way to do this with `jest.fn()` as you may see in the commit above.

File: typescript/src/AuctionMessageTranslator.ts [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+export class AuctionMessageTranslator {
+    private readonly listener: AuctionEventListener;
+
+    constructor(listener: AuctionEventListener) {
+        this.listener = listener;
+    }
+
+    public processMessage(message: string) {
+        if (message.includes('CLOSE')) {
+            this.listener.auctionClosed();
```
Comment (emilybache) [inline typescript/src/AuctionMessageTranslator.ts]: I think you have done a little too much here - should be just a comment that there is a bug here

Comment (ilandikov) [reply typescript/src/AuctionMessageTranslator.ts]: Indeed, fixed.

File: typescript/src/AuctionMessageTranslator.ts [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+        } else if (message.includes('PRICE')) {
+            const data: { [key: string]: string } = {};
```
Comment (emilybache) [inline typescript/src/AuctionMessageTranslator.ts]: I may not have done this consistently on all the language versions - but I think it's a good idea to extract a method here to get the data from the message - see here in the C# it's called ParseMessageData https://github.com/emilybache/Auction-TestDesign-Kata/blob/main/csharp/Auction-TestDesign-Kata/Auction/AuctionMessageTranslator.cs

**Comment by emilybache** [comment]
A comment on your solution that you linked to earlier in this thread. What you've done is a good use of the jest framework and is less work to implement than a hand-coded mock. However it feels more like a spy than a mock to me. The distinction is perhaps not worth the extra effort, but my understanding of JMock, the framework that Pryce and Freeman used in their book, is that mocks set expectations in advance - ie in the Arrange part of the test. They fail the test as soon as any expectation is breached. This gives you stacktraces and failures that happen in the 'act' step. The 'assert' part of the test is implicitly handled by the framework and often that part of the test is empty. With a spy you just record what happens and all the failures happen in the 'assert' part of the test. So spies are less proactive and fail more slowly. Often it doesn't matter, but sometimes it does. Mocks fail sooner and can help you find problems earlier.

File: typescript/src/AuctionMessageTranslator.ts (hunk 2 above)
Comment (ilandikov) [reply typescript/src/AuctionMessageTranslator.ts]: I did this in my solution =) Done here too now

**Comment by emilybache** [comment]
Thanks for the very prompt code updates!

**Comment by ilandikov** [comment]
> A comment on your solution that you linked to earlier in this thread. What you've done is a good use of the jest framework and is less work to implement than a hand-coded mock. However it feels more like a spy than a mock to me. The distinction is perhaps not worth the extra effort, but my understanding of JMock, the framework that Pryce and Freeman used in their book, is that mocks set expectations in advance - ie in the Arrange part of the test. They fail the test as soon as any expectation is breached. This gives you stacktraces and failures that happen in the 'act' step. The 'assert' part of the test is implicitly handled by the framework and often that part of the test is empty. With a spy you just record what happens and all the failures happen in the 'assert' part of the test. So spies are less proactive and fail more slowly. Often it doesn't matter, but sometimes it does. Mocks fail sooner and can help you find problems earlier.

Ok, thanks, I agree that there is a difference in the "how" here and how to call this (mock vs spy) where I totally agree with you. Do we agree that the both things are still pretty close in regards to getting the job done? (In this case, developing the interface)

**Comment by emilybache** [comment]
You can get design benefits when using a spy or a mock. The important part is to design the test first, including the expectations of what interactions will happen, before writing the implementation. Since the spy puts those at the end of the test you need to not give in to the temptation to write those assertions after the implementation of the act step is already built.

**Comment by ilandikov** [comment]
Of course =) Thanks!

**Pull request merged**